### PR TITLE
MSVC compile correctly. !!! NO TO MERGE!!!

### PR DESCRIPTION
**Pull request by skuater**
ONLY FOR REVIEW

-Plz dont merge yet , i need rewrite the grub code, now i made a dinamyc allocs to avoid compiler error with initialization from vars, but i dont insert code to free theses buffer, and i made without conditional compilation, i need add #ifdef to only do this change on msvc compiler.

### CODE REVIEW

File: binr/radare2/radare2.c [hunk 1]
```diff
@@ -2,7 +2,11 @@
 
 #define USE_THREADS 1
 #define UNCOLORIZE_NONTTY 0
-
+#ifdef _MSC_VER
```
Comment (radare) [inline binr/radare2/radare2.c]: this is not used anywhere, remove it

File: libr/asm/arch/arc/gnu/arc-ext.h [hunk 2]
```diff
@@ -84,14 +84,14 @@ enum ExtReadWrite
 
 /* constructor function */
 extern void build_ARC_extmap (void* text_bfd);
-
+#ifndef _MSC_VER
```
Comment (radare) [inline libr/asm/arch/arc/gnu/arc-ext.h]: what about #if 0

File: libr/asm/arch/avr/format.c [hunk 3]
```diff
@@ -27,7 +27,9 @@
 #include <string.h>
 #include <stdlib.h>
 #include "format.h"
-
+#ifdef _MSC_VER
+#define strcasecmp stricmp
```
Comment (radare) [inline libr/asm/arch/avr/format.c]: what about enhancing R_API const char *r_str_casestr(const char *a, const char *b) { ?

Comment (radare) [inline libr/asm/arch/avr/format.c]: errr. this strcasecmp thing should be used with a define in r_types.h like this

#if _MSC_VER
#define R_STRCASECMP stricmp
#else 
#define R_STRCASECMP strcasecmp
#endif 

or maybe we can just define static inline functions in r_util.h

File: libr/asm/arch/sparc/gnu/sparc-dis.c [hunk 4]
```diff
@@ -29,7 +29,15 @@
 #include "sysdep.h"
 #include "opcode/sparc.h"
 #include "dis-asm.h"
+#ifndef _MSC_VER
 #include "libiberty.h"
+#else
+#include <stdlib.h>
```
Comment (radare) [inline libr/asm/arch/sparc/gnu/sparc-dis.c]: why not put this code inside libiberty.h ?

File: libr/asm/arch/tricore/gnu/tricore-dis.c [hunk 5]
```diff
@@ -25,7 +25,15 @@ Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */
 #include "sysdep.h"
 #include "opcode/tricore.h"
 #include "dis-asm.h"
+#ifndef _MSC_VER
 #include "libiberty.h"
+#else
```
Comment (radare) [inline libr/asm/arch/tricore/gnu/tricore-dis.c]: you can remove this if you put the code in libibery.h

File: libr/asm/arch/xtensa/gnu/xtensa-dis.c [hunk 6]
```diff
@@ -26,10 +26,19 @@
 #include <string.h>
 #include "xtensa-isa.h"
 #include "ansidecl.h"
-#include "libiberty.h"
+#include <setjmp.h>
 #include "dis-asm.h"
+#ifndef _MSC_VER
+#include "libiberty.h"
```
Comment (radare) [inline libr/asm/arch/xtensa/gnu/xtensa-dis.c]: same here

File: libr/asm/arch/z80/z80asm.h [hunk 7]
```diff
@@ -24,7 +24,9 @@
 
 #include <stdio.h>
 #include <string.h>
+#ifndef _MSC_VER
 #include <strings.h>
```
Comment (radare) [inline libr/asm/arch/z80/z80asm.h]: this include is not necessary on linux either, remove it

File: libr/config/config.c [hunk 8]
```diff
@@ -2,7 +2,9 @@
 
 #include "r_config.h"
 #include "r_util.h" // r_str_hash, r_str_chop, ...
-
+#ifdef _MSC_VER
```
Comment (radare) [inline libr/config/config.c]: unnecessary if you put this in r_types.h or so

File: libr/include/msvc/getopt.h [hunk 9]
```diff
@@ -0,0 +1,72 @@
+#include <string.h>
```
Comment (radare) [inline libr/include/msvc/getopt.h]: not following r2 indent rules

File: libr/lang/lang.c [hunk 10]
```diff
@@ -13,6 +13,9 @@ R_LIB_VERSION(r_lang);
 #if __UNIX__
 #include "p/cpipe.c" // hardcoded
 #endif
+#ifdef _MSC_VER
```
Comment (radare) [inline libr/lang/lang.c]: unnecessary if moved into a .h

File: shlr/grub/fs/ext2.c [hunk 11]
```diff
@@ -385,7 +385,11 @@ grub_ext2_read_block (grub_fshelp_node_t node, grub_disk_addr_t fileblock)
 
   if (grub_le_to_cpu32(inode->flags) & EXT4_EXTENTS_FLAG)
     {
-      char buf[EXT2_BLOCK_SIZE(data)];
+#ifndef _MSC_VER
+	  char buf[EXT2_BLOCK_SIZE (data)];
```
Comment (radare) [inline shlr/grub/fs/ext2.c]: dont do this. this will allocate too much mem in the stack and probably crash. just replace grub_malloc for malloc(

File: shlr/grub/fs/ext2.c [hunk 12]
```diff
@@ -674,7 +678,11 @@ grub_ext2_iterate_dir (grub_fshelp_node_t dir,
 
       if (dirent.namelen != 0)
 	{
-	  char filename[dirent.namelen + 1];
+#ifndef _MSC_VER
+	  char filename[dirent.namelen + 1]; 
+#else
+	  char * filename = grub_malloc (dirent.namelen + 1);
```
Comment (radare) [inline shlr/grub/fs/ext2.c]: same here

File: shlr/grub/fs/jfs.c [hunk 13]
```diff
@@ -701,8 +720,11 @@ static grub_err_t
 grub_jfs_lookup_symlink (struct grub_jfs_data *data, int ino)
 {
   int size = grub_le_to_cpu64 (data->currinode.size);
+#ifndef _MSC_VER
   char symlink[size + 1];
-
+#else
+  char *symlink = grub_malloc(size + 1);
```
Comment (radare) [inline shlr/grub/fs/jfs.c]: use malloc, no ifdefs pls

File: shlr/grub/fs/jfs.c [hunk 14]
```diff
@@ -609,7 +624,11 @@ grub_jfs_read_file (struct grub_jfs_data *data,
 static grub_err_t
 grub_jfs_find_file (struct grub_jfs_data *data, const char *path)
 {
+#ifndef _MSC_VER
   char fpath[grub_strlen (path)];
+#else
+  char * fpath = grub_malloc(grub_strlen (path));
```
Comment (radare) [inline shlr/grub/fs/jfs.c]: same h

File: shlr/grub/kern/misc.c [hunk 15]
```diff
@@ -704,8 +704,12 @@ grub_vsnprintf_real (char *str, grub_size_t max_len, const char *fmt, va_list ar
 
 	  if (p > fmt)
 	    {
-	      char s[p - fmt + 1];
-	      grub_strncpy (s, fmt, p - fmt);
+#ifndef _MSC_VER
+		  char s[p - fmt + 1];
+#else
+		  char *s = grub_malloc(p - fmt + 1);
```
Comment (radare) [inline shlr/grub/kern/misc.c]: use malloc

File: shlr/grub/kern/partition.c [hunk 16]
```diff
@@ -213,8 +213,12 @@ grub_partition_get_name (const grub_partition_t partition)
     {
       /* Even on 64-bit machines this buffer is enough to hold
 	 longest number.  */
-      char buf[grub_strlen (part->partmap->name) + 25];
-      int strl;
+#ifndef _MSC_VER
+	  char buf[grub_strlen (part->partmap->name) + 25];
+#else
+	  char * buf = grub_malloc(grub_strlen (part->partmap->name) + 25);
```
Comment (radare) [inline shlr/grub/kern/partition.c]: no malloc

File: shlr/tcc/tccpp.c [hunk 17]
```diff
@@ -19,7 +19,9 @@
  */
 
 #include "tcc.h"
-
+#ifdef _MSC_VER
```
Comment (radare) [inline shlr/tcc/tccpp.c]: there's no problem to include this on linux, i prefer to avoid unnecessaries ifdefs

File: shlr/spp/spp.h [hunk 18]
```diff
@@ -28,10 +28,13 @@ int r_sys_setenv(const char *key, const char *value);
 extern int ifl;
 extern int echo[MAXIFL];
 extern int lineno;
-
 #ifndef DLL_LOCAL
+#ifdef _MSC_VER
```
Comment (radare) [inline shlr/spp/spp.h]: pls send a PR to spp repo for this

File: binr/radare2/radare2.c (hunk 1 above)
Comment (skuater) [inline binr/radare2/radare2.c]: what u mean? remove ifdef _MSC_VER or remove this and defines bellow?

Defines bellow are used to avoid missing headers form winsock into winbase.h/windows.h

File: libr/asm/arch/avr/format.c (hunk 3 above)
Comment (skuater) [inline libr/asm/arch/avr/format.c]: yes the problem are with files where no r_types.h/r_util.h inclusion how in this case. The solution is add int r_rtypes and add include to r_type.h in each file with stricmp/strcasecmp

File: libr/asm/arch/sparc/gnu/sparc-dis.c (hunk 4 above)
Comment (skuater) [inline libr/asm/arch/sparc/gnu/sparc-dis.c]: yeha just thinking about do it when revising for grub. is the next step

File: shlr/grub/fs/ext2.c (hunk 11 above)
Comment (skuater) [inline shlr/grub/fs/ext2.c]: grub_malloc is a wrapper for malloc. and the  char buf[EXT2_BLOCK_SIZE (data)]; is the original code form this file, i set the conditional compilation why else we need revise all code to made the free().

**Comment by radare** [comment]
what i want is to avoid 
> On 11 May 2017, at 23:37, SkUaTeR <notifications@github.com> wrote:
> 
> @skuater commented on this pull request.
> 
> In libr/asm/arch/avr/format.c <https://github.com/radare/radare2/pull/7476#discussion_r116106526>:
> 
> > @@ -27,7 +27,9 @@
>  #include <string.h>
>  #include <stdlib.h>
>  #include "format.h"
> -
> +#ifdef _MSC_VER
> +#define strcasecmp stricmp
> yes the problem are with files where no r_types.h/r_util.h inclusion how in this case. The solution is add int r_rtypes and add include to r_type.h in each file with stricmp/strcasecmp
> 
> 

i think for sanity we should “own” all this 3rd party code. and rewrite it using r2 style and remove all the code we dont need, etc.. not sure if we can even “upgrade” to grub2 this code without replacing it ocmpletely.

> 
> In libr/asm/arch/sparc/gnu/sparc-dis.c <https://github.com/radare/radare2/pull/7476#discussion_r116106686>:
> 
> >  #include "libiberty.h"
> +#else
> +#include <stdlib.h>
> yeha just thinking about do it when revising for grub. is the next step
> 
> yep

yep

> 
> In shlr/grub/fs/ext2.c <https://github.com/radare/radare2/pull/7476#discussion_r116108000>:
> 
> > @@ -385,7 +385,11 @@ grub_ext2_read_block (grub_fshelp_node_t node, grub_disk_addr_t fileblock)
>  
>    if (grub_le_to_cpu32(inode->flags) & EXT4_EXTENTS_FLAG)
>      {
> -      char buf[EXT2_BLOCK_SIZE(data)];
> +#ifndef _MSC_VER
> +	  char buf[EXT2_BLOCK_SIZE (data)];
> grub_malloc is a wrapper for malloc. and the char buf[EXT2_BLOCK_SIZE (data)]; is the original code form this file, i set the conditional compilation why else we need revise all code to made the free().
> 
> 
yep i know, we want to do the malloc way . always, dont make this conditional.

File: libr/asm/arch/include/libiberty.h [hunk 19]
```diff
@@ -308,7 +308,12 @@ extern double physmem_available (void);
    as new/delete and new[]/delete[].  */
 
 /* Scalar allocators.  */
-
+#ifdef _MSC_VER
+#include <stdlib.h>
```
Comment (radare) [inline libr/asm/arch/include/libiberty.h]: Linux doesnt have any of those xmalloc, xcalloc.. why is this necessary on windows?

File: libr/bin/mangling/cxx/libiberty.h [hunk 20]
```diff
@@ -333,7 +333,12 @@ extern unsigned int xcrc32 (const unsigned char *, int, unsigned int);
    as new/delete and new[]/delete[].  */
 
 /* Scalar allocators.  */
-
+#ifdef _MSC_VER
+#include <stdlib.h>
```
Comment (radare) [inline libr/bin/mangling/cxx/libiberty.h]: Same here

File: libr/core/cbin.c [hunk 21]
```diff
@@ -63,8 +63,7 @@ static void pair_str(const char *a, const char *b, int mode, int last) {
 }
 
 #define STR(x) (x)?(x):""
-// XXX - this may lead to conflicts with set by name
-static int r_core_bin_set_cur(RCore *core, RBinFile *binfile);
+R_API int r_core_bin_set_cur (RCore *core, RBinFile *binfile);
```
Comment (radare) [inline libr/core/cbin.c]: Change signature to return bool?

File: libr/include/msvc/getopt.h (hunk 9 above)
Comment (radare) [inline libr/include/msvc/getopt.h]: Run r2-indent or clean the syntax plz this is so ugly :p

File: shlr/grub/fs/ext2.c (hunk 11 above)
Comment (radare) [inline shlr/grub/fs/ext2.c]: I know. Just kill the old code and use malloc for all targetsz remove the ifdef
